File: erlang/src/htmlsanitizer.erl
%%% htmlsanitizer — clean HTML of constructs that can lead to XSS.
%%%
%%% A thin Erlang binding over the monorepo's ONE shared native engine
%%% (core/native/libhtmlsanitizer.so, compiled from pure Aether). No sanitizer
%%% logic lives here: every function marshals to an `aether_hs_embed_*` call
%%% through htmlsanitizer_nif.
%%%
%%%     {ok, S} = htmlsanitizer:new(),
%%%     <<"<div>Hello </div>">> =
%%%         htmlsanitizer:sanitize(S, <<"<div>Hello <script>x</script></div>">>),
%%%     ok = htmlsanitizer:close(S).
%%%
%%% This module's job is to turn the NIF's integer selectors into atoms, and
%%% its {ok, Binary} tuples into plain binaries for the common path. The
%%% *_r ("result") variants keep the tuple where a caller wants to distinguish
%%% a closed sanitizer from a legitimately empty result.
%%%
%%% NOTE ON CALLBACKS: this binding exposes NONE of the engine's hooks
%%% (on_removing_tag, on_filter_url, …). See README.md — a NIF cannot safely
%%% call back into the BEAM synchronously, so conformance checks 10 and 11 are
%%% skipped rather than faked.
-module(htmlsanitizer).

-export([new/0, close/1, is_closed/1,
         sanitize/2, sanitize/3, sanitize_r/3,
         sanitize_document/2, sanitize_document/3, sanitize_document_r/3,
         keep_child_nodes/1, set_keep_child_nodes/2,
         allow_data_attributes/1, set_allow_data_attributes/2,
         allow/3, disallow/3, is_allowed/3,
         clear/2, count/2, items/2, sorted_items/2,
         abi_version/0]).

-type sanitizer() :: htmlsanitizer_nif:sanitizer().
-export_type([sanitizer/0, list_name/0]).

%% The engine's six policy lists. These atoms map onto the ABI's integer
%% selectors, which are append-only and must never be renumbered.
-type list_name() :: tags
                   | attributes
                   | css_properties
                   | schemes
                   | classes
                   | uri_attributes.

%%------------------------------------------------------------------
%% Lifecycle
%%------------------------------------------------------------------

%% Create a sanitizer with the engine's secure defaults populated.
-spec new() -> {ok, sanitizer()} | {error, term()}.
new() -> htmlsanitizer_nif:new().

%% Release the native handle. Safe to call more than once. The resource's
%% destructor would do this anyway at GC time; close/1 makes it deterministic.
-spec close(sanitizer()) -> ok.
close(S) -> htmlsanitizer_nif:close(S).

-spec is_closed(sanitizer()) -> boolean().
is_closed(S) -> htmlsanitizer_nif:is_closed(S).

%%------------------------------------------------------------------
%% Sanitizing
%%------------------------------------------------------------------

%% Clean an HTML fragment with no base URL (relative URLs are not resolved).
-spec sanitize(sanitizer(), iodata()) -> binary().
sanitize(S, Html) -> sanitize(S, Html, <<>>).

%% Clean an HTML fragment, resolving relative URLs against BaseUrl.
%% Returns <<>> on a closed sanitizer; use sanitize_r/3 to tell the two apart.
-spec sanitize(sanitizer(), iodata(), iodata()) -> binary().
sanitize(S, Html, BaseUrl) -> unwrap(sanitize_r(S, Html, BaseUrl)).

-spec sanitize_r(sanitizer(), iodata(), iodata()) -> {ok, binary()} | {error, closed}.
sanitize_r(S, Html, BaseUrl) -> htmlsanitizer_nif:sanitize(S, Html, BaseUrl).

-spec sanitize_document(sanitizer(), iodata()) -> binary().
sanitize_document(S, Html) -> sanitize_document(S, Html, <<>>).

-spec sanitize_document(sanitizer(), iodata(), iodata()) -> binary().
sanitize_document(S, Html, BaseUrl) -> unwrap(sanitize_document_r(S, Html, BaseUrl)).

-spec sanitize_document_r(sanitizer(), iodata(), iodata()) -> {ok, binary()} | {error, closed}.
sanitize_document_r(S, Html, BaseUrl) -> htmlsanitizer_nif:sanitize_document(S, Html, BaseUrl).

unwrap({ok, Bin}) -> Bin;
unwrap({error, _}) -> <<>>.

%%------------------------------------------------------------------
%% Flags
%%------------------------------------------------------------------

%% Keep the children of a removed element instead of dropping the subtree.
-spec keep_child_nodes(sanitizer()) -> boolean().
keep_child_nodes(S) -> htmlsanitizer_nif:get_keep_child_nodes(S).

-spec set_keep_child_nodes(sanitizer(), boolean()) -> ok.
set_keep_child_nodes(S, On) when is_boolean(On) ->
    htmlsanitizer_nif:set_keep_child_nodes(S, On).

%% Let data-* attributes through without listing each one.
-spec allow_data_attributes(sanitizer()) -> boolean().
allow_data_attributes(S) -> htmlsanitizer_nif:get_allow_data_attributes(S).

-spec set_allow_data_attributes(sanitizer(), boolean()) -> ok.
set_allow_data_attributes(S, On) when is_boolean(On) ->
    htmlsanitizer_nif:set_allow_data_attributes(S, On).

%%------------------------------------------------------------------
%% Policy lists
%%------------------------------------------------------------------

%% ABI selectors — append only, never renumber (core/embed.ae).
-spec which(list_name()) -> 0..5.
which(tags)           -> 0;
which(attributes)     -> 1;
which(css_properties) -> 2;
which(schemes)        -> 3;
which(classes)        -> 4;
which(uri_attributes) -> 5.

%% Add an item (or a list of items) to a policy list.
-spec allow(sanitizer(), list_name(), iodata() | [iodata()]) -> boolean().
allow(S, List, Items) when is_list(Items), not is_integer(hd(Items)) ->
    lists:all(fun(I) -> allow(S, List, I) end, Items);
allow(S, List, Item) ->
    htmlsanitizer_nif:allow(S, which(List), Item).

%% Remove an item from a policy list (the "deny" direction).
-spec disallow(sanitizer(), list_name(), iodata() | [iodata()]) -> boolean().
disallow(S, List, Items) when is_list(Items), not is_integer(hd(Items)) ->
    lists:all(fun(I) -> disallow(S, List, I) end, Items);
disallow(S, List, Item) ->
    htmlsanitizer_nif:disallow(S, which(List), Item).

-spec is_allowed(sanitizer(), list_name(), iodata()) -> boolean().
is_allowed(S, List, Item) -> htmlsanitizer_nif:is_allowed(S, which(List), Item).

%% Empty a policy list — the "start from nothing" move for a caller who wants
%% a strict allow-list rather than the permissive defaults.
-spec clear(sanitizer(), list_name()) -> boolean().
clear(S, List) -> htmlsanitizer_nif:clear(S, which(List)).

-spec count(sanitizer(), list_name()) -> non_neg_integer().
count(S, List) -> htmlsanitizer_nif:count(S, which(List)).

%% Enumerate a policy list. Order is unspecified but stable between
%% mutations; use sorted_items/2 when a deterministic order matters.
-spec items(sanitizer(), list_name()) -> [binary()].
items(S, List) -> htmlsanitizer_nif:items(S, which(List)).

-spec sorted_items(sanitizer(), list_name()) -> [binary()].
sorted_items(S, List) -> lists:sort(items(S, List)).

%%------------------------------------------------------------------
%% Introspection
%%------------------------------------------------------------------

%% The engine's ABI revision.
-spec abi_version() -> non_neg_integer().
abi_version() -> htmlsanitizer_nif:abi_version().

File: elixir/lib/html_sanitizer.ex
defmodule HtmlSanitizer do
  @moduledoc """
  Clean HTML of constructs that can lead to Cross-Site Scripting (XSS).

  This is a thin Elixir surface over the monorepo's **canonical BEAM NIF**,
  which lives in `erlang/` and is compiled exactly once. There is no C source
  in this directory and no second `.so` — every function here `defdelegate`s
  to `:htmlsanitizer_nif`, the very same compiled module the Erlang and Gleam
  bindings load. One engine, one NIF, three languages.

  The engine itself (`core/native/libhtmlsanitizer.so`) is pure Aether. No
  sanitizer logic lives anywhere in this file: everything marshals to an
  `aether_hs_embed_*` call across the C ABI in `core/embed.ae`.

      iex> {:ok, s} = HtmlSanitizer.new()
      iex> HtmlSanitizer.sanitize(s, "<div>Hello <script>evil()</script></div>")
      "<div>Hello </div>"

  ## Finding the NIF

  `mix` does **not** honour `ERL_LIBS`, unlike `erl`, `escript` and `gleam`.
  So the app built by `erlang/.build.ae` is put on the code path explicitly —
  see `test/test_helper.exs`, which reads `$HTMLSANITIZER_BEAM_APP` and calls
  `Code.append_path/1`. That is the one wrinkle of consuming an OTP app that
  Mix did not build.

  ## Callbacks

  This binding exposes **none** of the engine's hooks, so conformance checks
  10 and 11 are not implemented. A NIF cannot synchronously call back into the
  BEAM. See `README.md` for the full reasoning — the surface is absent rather
  than faked.
  """

  @typedoc "An opaque handle to a native sanitizer (an `enif_resource`)."
  @opaque t :: reference()

  @typedoc """
  One of the engine's six policy lists.

  These map onto the ABI's integer selectors, which are append-only and must
  never be renumbered.
  """
  @type list_name ::
          :tags
          | :attributes
          | :css_properties
          | :schemes
          | :classes
          | :uri_attributes

  # ---- lifecycle ----

  @doc """
  Create a sanitizer with the engine's secure defaults populated.

  The handle is an `enif_resource`: the BEAM's GC releases the native handle
  when the last reference goes, so a dropped sanitizer leaks nothing.
  `close/1` makes that deterministic.
  """
  @spec new() :: {:ok, t()} | {:error, term()}
  defdelegate new(), to: :htmlsanitizer_nif

  @doc """
  Release the native handle. Idempotent.

  After this, `closed?/1` is `true` and `sanitize_r/3` returns
  `{:error, :closed}` rather than dereferencing a freed pointer.
  """
  @spec close(t()) :: :ok
  defdelegate close(sanitizer), to: :htmlsanitizer_nif

  @doc "Whether `close/1` has been called."
  @spec closed?(t()) :: boolean()
  defdelegate closed?(sanitizer), to: :htmlsanitizer_nif, as: :is_closed

  @doc """
  Run `fun` with a fresh sanitizer, closing it afterwards even on a raise.

  The Elixir-idiomatic bracket. Prefer it over hand-written new/close pairs.

      HtmlSanitizer.with_sanitizer(fn s ->
        HtmlSanitizer.sanitize(s, html)
      end)
  """
  @spec with_sanitizer((t() -> result)) :: result when result: var
  def with_sanitizer(fun) when is_function(fun, 1) do
    {:ok, s} = new()

    try do
      fun.(s)
    after
      close(s)
    end
  end

  # ---- sanitizing ----

  @doc """
  Clean an HTML fragment.

  `base_url` resolves relative URLs; pass `""` (the default) for no
  resolution. Input is `iodata`, so an unflattened iolist is fine.

  Returns `""` on a closed sanitizer — use `sanitize_r/3` when you need that
  distinguished from a legitimately empty result.

      iex> HtmlSanitizer.with_sanitizer(fn s ->
      ...>   HtmlSanitizer.sanitize(s, ~s(<img src="logo.png">), "https://example.com")
      ...> end)
      ~s(<img src="https://example.com/logo.png">)
  """
  @spec sanitize(t(), iodata(), iodata()) :: binary()
  def sanitize(sanitizer, html, base_url \\ ""), do: unwrap(sanitize_r(sanitizer, html, base_url))

  @doc "Like `sanitize/3` but reports `{:error, :closed}` instead of `\"\"`."
  @spec sanitize_r(t(), iodata(), iodata()) :: {:ok, binary()} | {:error, :closed}
  defdelegate sanitize_r(sanitizer, html, base_url), to: :htmlsanitizer_nif, as: :sanitize

  @doc "Clean a whole HTML document."
  @spec sanitize_document(t(), iodata(), iodata()) :: binary()
  def sanitize_document(sanitizer, html, base_url \\ ""),
    do: unwrap(sanitize_document_r(sanitizer, html, base_url))

  @doc "Like `sanitize_document/3` but reports `{:error, :closed}`."
  @spec sanitize_document_r(t(), iodata(), iodata()) :: {:ok, binary()} | {:error, :closed}
  defdelegate sanitize_document_r(sanitizer, html, base_url),
    to: :htmlsanitizer_nif,
    as: :sanitize_document

  defp unwrap({:ok, bin}), do: bin
  defp unwrap({:error, _}), do: ""

  # ---- flags ----

  @doc "Whether children of a removed element are kept."
  @spec keep_child_nodes(t()) :: boolean()
  defdelegate keep_child_nodes(sanitizer), to: :htmlsanitizer_nif, as: :get_keep_child_nodes

  @doc "Keep the children of a removed element instead of dropping the subtree."
  @spec set_keep_child_nodes(t(), boolean()) :: :ok
  defdelegate set_keep_child_nodes(sanitizer, on), to: :htmlsanitizer_nif

  @doc "Whether `data-*` attributes pass without being listed."
  @spec allow_data_attributes(t()) :: boolean()
  defdelegate allow_data_attributes(sanitizer),
    to: :htmlsanitizer_nif,
    as: :get_allow_data_attributes

  @doc "Let `data-*` attributes through without listing each one."
  @spec set_allow_data_attributes(t(), boolean()) :: :ok
  defdelegate set_allow_data_attributes(sanitizer, on), to: :htmlsanitizer_nif

  # ---- policy lists ----
  #
  # The ABI takes an integer selector; Elixir callers name the list with an
  # atom and never see the number. These constants are ABI — append only,
  # never renumber (core/embed.ae).

  @which %{
    tags: 0,
    attributes: 1,
    css_properties: 2,
    schemes: 3,
    classes: 4,
    uri_attributes: 5
  }

  defp which(name) when is_map_key(@which, name), do: Map.fetch!(@which, name)

  @doc """
  Add one item, or a list of items, to a policy list.

      HtmlSanitizer.allow(s, :tags, "my-widget")
      HtmlSanitizer.allow(s, :tags, ["one", "two"])
  """
  @spec allow(t(), list_name(), iodata() | [iodata()]) :: boolean()
  def allow(sanitizer, list, items) when is_list(items),
    do: Enum.all?(items, &allow(sanitizer, list, &1))

  def allow(sanitizer, list, item),
    do: :htmlsanitizer_nif.allow(sanitizer, which(list), item)

  @doc "Remove an item from a policy list (the \"deny\" direction)."
  @spec disallow(t(), list_name(), iodata() | [iodata()]) :: boolean()
  def disallow(sanitizer, list, items) when is_list(items),
    do: Enum.all?(items, &disallow(sanitizer, list, &1))

  def disallow(sanitizer, list, item),
    do: :htmlsanitizer_nif.disallow(sanitizer, which(list), item)

  @doc "Whether an item is currently in a policy list."
  @spec allowed?(t(), list_name(), iodata()) :: boolean()
  def allowed?(sanitizer, list, item),
    do: :htmlsanitizer_nif.is_allowed(sanitizer, which(list), item)

  @doc """
  Empty a policy list.

  The "start from nothing" move for a caller who wants a strict allow-list
  rather than the engine's permissive defaults.
  """
  @spec clear(t(), list_name()) :: boolean()
  def clear(sanitizer, list), do: :htmlsanitizer_nif.clear(sanitizer, which(list))

  @doc "How many entries a policy list has."
  @spec count(t(), list_name()) :: non_neg_integer()
  def count(sanitizer, list), do: :htmlsanitizer_nif.count(sanitizer, which(list))

  @doc """
  Enumerate a policy list.

  Order is unspecified but stable between mutations; use `sorted_items/2` when
  determinism matters.
  """
  @spec items(t(), list_name()) :: [binary()]
  def items(sanitizer, list), do: :htmlsanitizer_nif.items(sanitizer, which(list))

  @doc "`items/2`, sorted."
  @spec sorted_items(t(), list_name()) :: [binary()]
  def sorted_items(sanitizer, list), do: Enum.sort(items(sanitizer, list))

  # ---- introspection ----

  @doc "The engine's ABI revision."
  @spec abi_version() :: non_neg_integer()
  defdelegate abi_version(), to: :htmlsanitizer_nif
end
